### ISSUE TITLE
Unprocessable Entity (422)

### ISSUE THREAD

**Issue by spraguep** [restlet/restlet-framework-java#1018]
When upgrading from Restlet JSE 2.2.1 to 2.3.0 I am getting an error. The problem has something to do with deserializing a "text/plain" 500 error into a StatusInfo object.

This is the stacktrace:
```
Unprocessable Entity (422) - The server understands the content type of the request entity and the syntax of the request entity is correct but was unable to process the contained instructions
	at org.restlet.resource.Resource.toObject(Resource.java:893)
	at org.restlet.engine.resource.ClientInvocationHandler.invoke(ClientInvocationHandler.java:310)
	at com.sun.proxy.$Proxy7.doGet(Unknown Source)
	at restletbugtest.BugTest.test(BugTest.java:40)
Caused by: java.io.IOException: Unable to create the Object representation
	at org.restlet.engine.converter.DefaultConverter.toObject(DefaultConverter.java:282)
	at org.restlet.service.ConverterService.toObject(ConverterService.java:229)
	at org.restlet.resource.Resource.toObject(Resource.java:889)
	... 36 more
Caused by: java.lang.IllegalArgumentException: The serialized representation must have this media type: application/x-java-serialized-object or this one: application/x-java-serialized-object+xml
	at org.restlet.representation.ObjectRepresentation.<init>(ObjectRepresentation.java:221)
	at org.restlet.representation.ObjectRepresentation.<init>(ObjectRepresentation.java:123)
	at org.restlet.representation.ObjectRepresentation.<init>(ObjectRepresentation.java:104)
	at org.restlet.engine.converter.DefaultConverter.toObject(DefaultConverter.java:279)
	... 38 more
```

Here is the pom file:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>restletbugtest</groupId>
    <artifactId>restletbugtest</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <packaging>jar</packaging>
    
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        
        <restlet.version>2.3.0</restlet.version>
        
        <!-- no problem with this version of restlet
        <restlet.version>2.2.1</restlet.version>
        -->
    </properties>
    
    <dependencies>
        <dependency>
            <groupId>org.restlet.jse</groupId>
            <artifactId>org.restlet</artifactId>
            <version>${restlet.version}</version>
        </dependency>
        <dependency>
            <groupId>org.restlet.jse</groupId>
            <artifactId>org.restlet.ext.json</artifactId>
            <version>${restlet.version}</version>
        </dependency>
        <dependency>
            <groupId>org.restlet.jse</groupId>
            <artifactId>org.restlet.ext.jackson</artifactId>
            <version>${restlet.version}</version>
        </dependency>
        <dependency>
            <groupId>org.restlet.jse</groupId>
            <artifactId>org.restlet.ext.slf4j</artifactId>
            <version>${restlet.version}</version>
        </dependency>
        <dependency>
            <groupId>org.restlet.jse</groupId>
            <artifactId>org.restlet.ext.httpclient</artifactId>
            <version>${restlet.version}</version>
        </dependency>
        
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>6.8.8</version>
            <scope>test</scope>
            <type>jar</type>
        </dependency>
    </dependencies>

</project>
```

Here is the test that reproduces the problem:
```java
package restletbugtest;

import java.io.IOException;
import java.net.ServerSocket;
import org.restlet.Application;
import org.restlet.Component;
import org.restlet.Restlet;
import org.restlet.data.Protocol;
import org.restlet.data.Status;
import org.restlet.resource.ClientProxy;
import org.restlet.resource.ClientResource;
import org.restlet.resource.Get;
import org.restlet.resource.ResourceException;
import org.restlet.resource.ServerResource;
import org.restlet.routing.Router;
import org.testng.Assert;
import org.testng.annotations.Test;

/**
 * Bug Test.
 */
public class BugTest {

    /**
     * Bug Test.
     *
     * @throws Exception thrown on error
     */
    @Test
    public void test() throws Exception {
        final int port = findFreePort();

        final Component mailServer = new Component();
        mailServer.getServers().add(Protocol.HTTP, port);
        mailServer.getDefaultHost().attach(new MyApplication());
        mailServer.start();

        final FailureResource resource = ClientResource.create("http://localhost:" + port + "/fail", FailureResource.class);
        try {
            resource.doGet();
            Assert.fail("expected exception not thrown");
        } catch (ResourceException e) {
            if (e.getStatus().getCode() == 500) {
                Assert.assertEquals("failure-response", ((ClientProxy) resource).getClientResource().getResponse().getEntityAsText());
            } else {
                throw e;
            }
        }
    }

    /**
     * Restlet application.
     */
    public static class MyApplication extends Application {

        /**
         * Creates a root Router to dispatch call to server resources.
         *
         * @return the root restlet
         */
        @Override
        public Restlet createInboundRoot() {
            Router router = new Router(getContext());
            router.attach("/fail", ServerFailureResource.class);
            return router;
        }
    }

    /**
     * Server side resource implementing the {@link FailureResource} interface.
     */
    public static class ServerFailureResource extends ServerResource implements FailureResource {
        @Override
        public String doGet() {
            setStatus(Status.SERVER_ERROR_INTERNAL, "FAIL");
            return "failure-response";
        }
    }

    /**
     * Restlet client interface used to create a dynamic proxy.
     */
    public static interface FailureResource {

        /**
         * Failure GET.
         *
         * @return failure response
         */
        @Get("txt")
        String doGet();
    }

    /**
     * Returns an available port number.
     *
     * @return a port number
     * @throws IOException thrown when unable to find free port
     */
    public static int findFreePort() throws IOException {
        final ServerSocket server = new ServerSocket(0);
        final int port = server.getLocalPort();
        server.close();
        return port;
    }
}
```

My system information:
```
Apache Maven 3.2.3 (33f8c3e1027c3ddde99d3cdebad2656a31e8fdf4; 2014-08-11T16:58:10-04:00)
Maven home: /Users/psprague/Apps/maven
Java version: 1.8.0_25, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_25.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.9.5", arch: "x86_64", family: "mac"
```

**Comment by thboileau** [comment]
many thanks @spraguep for providing a sample code. that helps a lot.

**Comment by thboileau** [comment]
linked with #1004

**Comment by thboileau** [comment]
I think the issue is located in the client code which infers too quickly that the entity of the error response is a serialized instance of StatusInfo.
I don't have a fix right now, we are having some thoughts about it.

Just in order to give some explanations.
The `StatusInfo` bean has been set up in order to furnish automatic serialization of exception.
In the previous releases of Restlet Framework, the error responses contained an HTML auto-generated entity, unless the user of RF generates his own error entity.
We decided to change this as follow, in case an exception is thrown by your resource's method and caught by the StatusService:

1. the framework is now able to serialize subclasses of Exception and provide the right response status using the [`Status` annotation](http://restlet.com/technical-resources/restlet-framework/javadocs/2.3/jse/api/org/restlet/resource/Status.html)
2. if not, a [`StatusInfo` instance](http://restlet.com/technical-resources/restlet-framework/javadocs/2.3/jse/engine/org/restlet/engine/application/StatusInfo.html) is serialized

Which leads us to update the client code by trying to deserialize an Exception, or a StatusInfo bean.
Obviously, this works only if all errors responses are handled by the framework, which is not the case. We have to take into account the cases where you want to provide your own representations.

**Comment by thboileau** [comment]
Hi @spraguep , we've removed the automatic deserialization of StatusInfo bean.
The fix will be provided by RF 2.3.1 coming very soon.
